payments/models.py

```diff
@@ -185,6 +243,39 @@ def get_success_url(self) -> str:
     def get_process_url(self) -> str:
         return reverse("process_payment", kwargs={"token": self.token})
 
+    def get_payment_url(self) -> str:
+        """
+        Get the url the view that handles the payment
+        (payment_details() in documentation)
+        For now used only by PayU provider to redirect users back to CVV2 form
+        """
+        raise NotImplementedError
+
+    def get_subscription(self) -> BaseSubscription | None:
+        """
+        Returns subscription object associated with this payment
+        or None if the payment is not recurring
+        """
+        return None
+
+    def is_recurring(self) -> bool:
+        return self.get_subscription() is not None
+
+    def autocomplete_with_subscription(self):
+        """
+        Complete the payment with subscription
+
+        If the provider uses workflow such that the payments are initiated from
+        implementer's side.
+        Call this function right before the subscription end to
+        make a new subscription payment.
+
+        Throws RedirectNeeded if there is problem with the payment
+        that needs to be solved by user
+        """
+        provider = provider_factory(self.variant)
+        provider.autocomplete_with_subscription(self)
```

Comment:
Oh, and... Would it make sense to check if the payment has an allowed `status` (`WAITING`) only?